### PR TITLE
Fix #1051 - Remember if we made changes that need to be published on refresh

### LINKED ISSUE #1051: Unable to publish when you reload the page
STR:
* Open a project
* Publish it
* Make some changes (and wait till it saves)
* Click on the Publish button at the top and the dialog says that you need to "Update published version" (don't click the button)
* Refresh the page
* Click on the Publish button at the top and you no longer see "Update published version". In fact, the only way to publish your changes is to unpublish an re-publish again

publish

### PR DESCRIPTION
**Pull request by gideonthomas**

### CODE REVIEW

File: public/editor/scripts/editor/js/fc/publisher.js [hunk 1]
```diff
@@ -66,10 +68,24 @@ define(function(require) {
...
+        publisher.enable();
+        publisher.handlers.unpublishedChangesPrompt();
+
+        if(publisher.needsUpdate) {
+          return;
+        }
+
+        Project.publishNeedsUpdate(true, function(err) {
+          if(err) {
+            console.error("[Bramble] Failed to set the publishNeedsUpdate flag after a file change: ", err);
```
Comment (humphd) [inline public/editor/scripts/editor/js/fc/publisher.js]: s/Thimble/Bramble/

File: public/editor/scripts/editor/js/fc/publisher.js [hunk 2]
```diff
@@ -82,6 +98,20 @@ define(function(require) {
     }
 
     dialog.buttons.publish.on("click", publisher.handlers.publish);
+
+    // Were there any files that were updated and not published?
+    Project.doesPublishNeedUpdate(function(err, needsUpdate) {
+      if(err) {
+        console.error("[Bramble] Failed to get the publishNeedsUpdate flag while loading the publish dialog: ", err);
```
Comment (humphd) [inline public/editor/scripts/editor/js/fc/publisher.js]: s/Bramble/Thimble/

File: public/editor/scripts/editor/js/fc/publisher.js [hunk 3]
```diff
@@ -115,6 +145,14 @@ define(function(require) {
         }
 
         publisher.updateDialog(project.link, true);
+        Project.publishNeedsUpdate(false, function(err) {
+          if(err) {
+            console.error("[Bramble] Failed to set the publishNeedsUpdate flag after publishing with: ", err);
```
Comment (humphd) [inline public/editor/scripts/editor/js/fc/publisher.js]: Same.  Just do this change everywhere.

File: public/editor/scripts/editor/js/fc/publisher.js [hunk 4]
```diff
@@ -173,6 +211,15 @@ define(function(require) {
       buttons.parent.removeClass("hide");
 
       publisher.updateDialog("");
+
+      Project.publishNeedsUpdate(false, function(err) {
+        if(err) {
+          console.error("[Bramble] Failed to set the publishNeedsUpdate flag after unpublishing with: ", err);
+          return;
```
Comment (humphd) [inline public/editor/scripts/editor/js/fc/publisher.js]: Sometimes you `console.error` and then return, other times you let it run through.  Should you always do the same?

Comment (gideonthomas) [inline public/editor/scripts/editor/js/fc/publisher.js]: In other cases I don't have code running after the if statement so I don't need to return

Comment (gideonthomas) [inline public/editor/scripts/editor/js/fc/publisher.js]: Or I'm wrong

File: public/editor/scripts/editor/js/fc/publisher.js [hunk 5]
```diff
@@ -66,10 +68,24 @@ define(function(require) {
...
+        publisher.handlers.unpublishedChangesPrompt();
+
+        if(publisher.needsUpdate) {
+          return;
+        }
+
+        Project.publishNeedsUpdate(true, function(err) {
+          if(err) {
+            console.error("[Bramble] Failed to set the publishNeedsUpdate flag after a file change: ", err);
+          }
```
Comment (gideonthomas) [inline public/editor/scripts/editor/js/fc/publisher.js]: Add a return here

File: public/editor/scripts/project/project.js [hunk 6]
```diff
@@ -91,6 +91,18 @@ define(function(require) {
...
 
+  // Sets a flag on the project root that indicates whether we need to update
+  // the published version of this project or not
+  function publishNeedsUpdate(value, callback) {
+    Metadata.setPublishNeedsUpdate(getRoot(), value, callback);
+  }
+
+  // Gets the flag from the project root that indicates whether we need to
+  // update the published version of this project or not
+  function doesPublishNeedUpdate(callback) {
```
Comment (humphd) [inline public/editor/scripts/project/project.js]: I would call this `getPublishNeedsUpdate` (also note the `s` on `Needs` that you missed).

**Comment by humphd** [comment]
Generally this looks good, with my comments addressed.  I also filed https://github.com/mozilla/thimble.webmaker.org/issues/1092, which is solving a similar issue for file uploads.

r=me with these things fixed and travis.

**Pull request merged**